ynr/apps/sopn_parsing/management/commands/sopn_parsing_aws_textract.py

```diff
@@ -75,64 +81,104 @@ def get_job_results(self, job_id):
-                                            for candidate, description in zip(
-                                                candidates, descriptions
-                                            ):
-                                                print(
-                                                    f"{candidate}: {description}"
-                                                )
+    def map_blocks(self, blocks, block_type):
+        return {
+            block["Id"]: block
+            for block in blocks
+            if block["BlockType"] == block_type
+        }
+
+    def get_children_ids(self, block):
+        for rels in block.get("Relationships", []):
+            if rels["Type"] == "CHILD":
+                yield from rels["Ids"]
+
+    def transform_response_into_df(self, textract_result):
+        """This step is to transform the raw data into a pandas dataframe so
+        that we can then weave it into the existing parsing process.
+        before we try to match parties."""
+        # with credit to https://maxhalford.github.io/blog/textract-table-to-pandas/
+        # keep an eye on this as it may be useful: https://pypi.org/project/amazon-textract-response-parser/
+        # for now, we're just going to use the raw data
+        import pandas as pd
+
+        blocks = textract_result["Blocks"]
+        tables = self.map_blocks(blocks, "TABLE")
+        cells = self.map_blocks(blocks, "CELL")
+        words = self.map_blocks(blocks, "WORD")
+        selections = self.map_blocks(blocks, "SELECTION_ELEMENT")
+        lines = self.map_blocks(blocks, "LINE")
+
+        dataframes = []
+
+        if tables:
+            # if there are tables, we know this is a structured document
+            for table in tables.values():
+                # Determine all the cells that belong to this table
+                table_cells = [
+                    cells[cell_id] for cell_id in self.get_children_ids(table)
+                ]
+
+                # Determine the table's number of rows and columns
+                n_rows = max(cell["RowIndex"] for cell in table_cells)
+                n_cols = max(cell["ColumnIndex"] for cell in table_cells)
+                content = [[None for _ in range(n_cols)] for _ in range(n_rows)]
+
+                # Fill in each cell
+                for cell in table_cells:
+                    cell_contents = [
+                        words[child_id]["Text"]
+                        if child_id in words
+                        else selections[child_id]["SelectionStatus"]
+                        for child_id in self.get_children_ids(cell)
+                    ]
+                    i = cell["RowIndex"] - 1
+                    j = cell["ColumnIndex"] - 1
+                    content[i][j] = " ".join(cell_contents)
+
+                dataframe = pd.DataFrame(content[1:], columns=content[0])
+                dataframes.append(dataframe)
```

Comment:
@symroe This was the scenario I mentioned on standup. If there are no table blocks found in the Textract result, but there are lines, we join and save as a string before converting to a df, but the result may prove to challenging to match headers, etc.